lib/src/merged_tree.rs

```diff
@@ -828,27 +807,10 @@ impl Iterator for TreeDiffIterator<'_> {
                     }
                 };
                 let subdir =
-                    TreeDiffDirItem::from_trees(&path, &before_tree, &after_tree, self.matcher);
-                self.stack.push(TreeDiffItem::Dir(subdir));
-                self.stack.len() - 1
-            } else {
-                self.stack.len()
+                    TreeDiffDir::from_trees(&path, &before_tree, &after_tree, self.matcher);
+                self.stack.push(subdir);
             };
-            if !tree_before && tree_after {
-                if before.is_present() {
-                    return Some(TreeDiffEntry {
-                        path,
-                        values: Ok((before, Merge::absent())),
-                    });
-                }
-            } else if tree_before && !tree_after {
-                if after.is_present() {
-                    self.stack.insert(
-                        post_subdir,
-                        TreeDiffItem::File(path, Merge::absent(), after),
-                    );
-                }
-            } else if !tree_before && !tree_after {
+            if before.is_file_like() || after.is_file_like() {
```

Comment:
(No action requested)

It might be better to let caller filter tree->tree diff entries. We might want to add `jj diff --name-only --include-directory` or something for completion of path prefixes.